azkaban-common/src/main/java/azkaban/executor/AbstractExecutorManagerAdapter.java

```diff
@@ -599,40 +601,45 @@ protected void handleCancelFlow(ExecutionReference executionReference,
         executionReference.getDispatchMethod() == DispatchMethod.CONTAINERIZED ? Status.KILLED :
             Status.FAILED;
     if (!isExecutionReachable(executionReference, userId)) {
+      final String finalizingReason = "Cancel action has been called but the flow is unreachable.";
       logger.info("Finalizing executable flow as execution is unreachable: " + executionReference
           .getExecId());
-      ExecutionControllerUtils.finalizeFlow(this, this.projectManager, this.executorLoader,
-          this.alerterHolder,
-          executableFlow, "Cancel action has been called but the flow is unreachable.", null,
+      handleCancelFlowManually(executionReference, executableFlow, finalizingReason, null,
           finalizingStatus);
-      // Killed flow events can only be sent out if callWithReferenceByUser completed successfully
-      // so we need to manually send one here.
-      this.fireEventListeners(Event.create(executableFlow,
-          EventType.FLOW_FINISHED, new EventData(executableFlow)));
-      // Throwing exception to make the reason appear on the UI.
-      throw new ExecutorManagerException("Flow execution is unreachable. Finalizing the flow.");
     }
 
     try {
       this.apiGateway.callWithReferenceByUser(executionReference, ConnectorParams.CANCEL_ACTION,
           userId);
     } catch (Exception e) {
+      final String finalizingReason = "Unable to gracefully kill the flow execution or flow is "
+          + "unreachable.";
       logger
           .error("Exception occurred while cancelling flow: " + executionReference.getExecId(), e);
-      logger.info("Finalizing executable flow: " + executionReference.getExecId());
-      final String finalizingReason = "Unable to gracefully kill the flow execution.";
-      ExecutionControllerUtils.finalizeFlow(this, this.projectManager, this.executorLoader,
-          this.alerterHolder, executableFlow, finalizingReason, e, finalizingStatus);
-      // Killed flow events can only be sent out if callWithReferenceByUser completed successfully
-      // so we need to manually send one here.
-      this.fireEventListeners(Event.create(executableFlow,
-          EventType.FLOW_FINISHED, new EventData(executableFlow)));
-      // Throwing exception to make the reason appear on the UI.
-      throw new ExecutorManagerException(finalizingReason
-          + " Finalizing the flow.");
+      handleCancelFlowManually(executionReference, executableFlow, finalizingReason, e,
+          finalizingStatus);
     }
   }
 
+  private void handleCancelFlowManually(ExecutionReference executionReference,
```

Comment:
Let's have comments for each extracted method. 
explanation what this method does and in which case these method should be used.
for example cancelFlowManually should be called when some exception hits and flow can not be gracefully killed. 
(BTW should we be consistent to use Kill or Cancel? it confused me a bit at first)

Reply:
Good point! Done.